bug: KISS does not work after reboot, green circle is rotating but i have to restart KISS to make it work

* app version 3.7.5 (from fdroid)
* android version P
* device used Xiaomi Redmi Note 5 Pro (whyred) and Xiaomi Mi Max 3 (nitrogen)
* custom settings that might have an impact: no idea (tried with and without the "root option" and with and without the "widget option") nothing changed

I use on both phones custom ROMs based on aosp (havoc os and cosp).

I had this happen to me last week.

I was able to restart (advanced settings, restart), and it was fine.

Here are the logs for me:

```
2019-02-14 01:57:39.943 15706-15706/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: fr.neamar.kiss, PID: 15706
    java.lang.RuntimeException: Unable to start activity ComponentInfo{fr.neamar.kiss/fr.neamar.kiss.MainActivity}: java.lang.ClassCastException: android.app.Application cannot be cast to fr.neamar.kiss.KissApplication
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2951)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3086)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1816)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6718)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
     Caused by: java.lang.ClassCastException: android.app.Application cannot be cast to fr.neamar.kiss.KissApplication
        at fr.neamar.kiss.KissApplication.getApplication(KissApplication.java:18)
        at fr.neamar.kiss.MainActivity.onCreate(MainActivity.java:153)
        at android.app.Activity.performCreate(Activity.java:7144)
        at android.app.Activity.performCreate(Activity.java:7135)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1271)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2931)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3086) 
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1816) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.app.ActivityThread.main(ActivityThread.java:6718) 
        at java.lang.reflect.Method.invoke(Native Method) 
```

It's a `RuntimeException`, but I didn't get an error message -- just a rotating circle.

Not sure if this was at fault, but I don't have many kiss entries
```
02-19 14:31:27.438   948  1317 W ActivityManager: Background start not allowed: service Intent { cmp=fr.neamar.[01;31m[Kkiss[m[K/.dataprovider.ContactsProvider } to fr.neamar.kiss/.dataprovider.ContactsProvider from pid=32005 uid=10141 pkg=fr.neamar.kiss startFg?=false
02-19 14:31:27.439 32005 32005 W System.err: java.lang.ClassNotFoundException: fr.neamar.[01;31m[Kkiss[m[K.dataprovider.PhoneProvider
02-19 14:31:27.440 32005 32005 W System.err: 	at fr.neamar.[01;31m[Kkiss[m[K.DataHandler.providerName2Intent(DataHandler.java:139)
02-19 14:31:27.440 32005 32005 W System.err: 	at fr.neamar.[01;31m[Kkiss[m[K.DataHandler.connectToProvider(DataHandler.java:161)
02-19 14:31:27.440 32005 32005 W System.err: 	at fr.neamar.[01;31m[Kkiss[m[K.DataHandler.<init>(DataHandler.java:93)
02-19 14:31:27.440 32005 32005 W System.err: 	at fr.neamar.[01;31m[Kkiss[m[K.KissApplication.getDataHandler(KissApplication.java:23)
02-19 14:31:27.440 32005 32005 W System.err: 	at fr.neamar.[01;31m[Kkiss[m[K.broadcast.IncomingCallHandler.onReceive(IncomingCallHandler.java:24)
02-19 14:31:27.441 32005 32005 W System.err: Caused by: java.lang.ClassNotFoundException: Didn't find class "fr.neamar.[01;31m[Kkiss[m[K.dataprovider.PhoneProvider" on path: DexPathList[[zip file "/data/app/fr.neamar.kiss-Gab5SFRMN7I3V0I7JQgT3A==/base.apk"],nativeLibraryDirectories=[/data/app/fr.neamar.kiss-Gab5SFRMN7I3V0I7JQgT3A==/lib/arm64, /system/lib64, /vendor/lib64]]
``` 

Ignore the cosmetic f-up, my logs output like that.

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -7,8 +7,8 @@ android {
         applicationId 'fr.neamar.kiss'
         minSdkVersion 15
         targetSdkVersion 28
-        versionCode 147
-        versionName "3.7.7"
+        versionCode 148
+        versionName "3.7.8"
         testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
     }
     buildTypes {
@@ -24,7 +24,7 @@ dependencies {
     androidTestImplementation 'androidx.test:runner:1.1.1'
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.1.1'
     androidTestImplementation 'androidx.test:rules:1.1.1'
-    implementation 'androidx.annotation:annotation:1.0.1'
+    implementation 'androidx.annotation:annotation:1.0.2'
 }
 
 buildscript {

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -7,8 +7,8 @@ android {
         applicationId 'fr.neamar.kiss'
         minSdkVersion 15
         targetSdkVersion 28
-        versionCode 146
-        versionName "3.7.6"
+        versionCode 147
+        versionName "3.7.7"
         testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
     }
     buildTypes {

--- a/app/src/main/java/fr/neamar/kiss/DataHandler.java
+++ b/app/src/main/java/fr/neamar/kiss/DataHandler.java
@@ -1,5 +1,6 @@
 package fr.neamar.kiss;
 
+import android.app.ActivityManager;
 import android.content.BroadcastReceiver;
 import android.content.ComponentName;
 import android.content.Context;
@@ -8,13 +9,11 @@ import android.content.IntentFilter;
 import android.content.ServiceConnection;
 import android.content.SharedPreferences;
 import android.graphics.Bitmap.CompressFormat;
-import android.hardware.display.DisplayManager;
 import android.os.Build;
 import android.os.IBinder;
 import android.preference.PreferenceManager;
 import android.text.TextUtils;
 import android.util.Log;
-import android.view.Display;
 
 import java.io.ByteArrayOutputStream;
 import java.util.ArrayList;
@@ -84,7 +83,8 @@ public class DataHandler extends BroadcastReceiver
             // However... KISS is a launcher, and when your device reboots, Android starts the launcher in the background
             // This is obviously an issue, so we need to wait until screen is really on to do something
             // https://github.com/Neamar/KISS/issues/1154
-            if (screenIsOn()) {
+            if (isScreenOn()) {
+                Log.v(TAG, "Screen is on, services can start normally");
                 init();
             }
             else {
@@ -682,14 +682,10 @@ public class DataHandler extends BroadcastReceiver
     }
 
     @RequiresApi(api = Build.VERSION_CODES.P)
-    private boolean screenIsOn() {
-        DisplayManager dm = (DisplayManager) context.getSystemService(Context.DISPLAY_SERVICE);
-        for (Display display : dm.getDisplays()) {
-            if (display.getState() == Display.STATE_ON) {
-                return true;
-            }
-        }
-        return false;
+    private boolean isScreenOn() {
+        ActivityManager.RunningAppProcessInfo appProcessInfo = new ActivityManager.RunningAppProcessInfo();
+        ActivityManager.getMyMemoryState(appProcessInfo);
+        return (appProcessInfo.importance == ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND || appProcessInfo.importance == ActivityManager.RunningAppProcessInfo.IMPORTANCE_VISIBLE);
     }
 
     public TagsHandler getTagsHandler() {

--- a/app/src/main/java/fr/neamar/kiss/DataHandler.java
+++ b/app/src/main/java/fr/neamar/kiss/DataHandler.java
@@ -184,19 +184,23 @@ public class DataHandler extends BroadcastReceiver
                 return;
             }
 
+            // Add a receiver to get notified next time the screen is on
+            // or next time the users successfully dismisses his lock screen
             IntentFilter intentFilter = new IntentFilter();
             intentFilter.addAction(Intent.ACTION_SCREEN_ON);
             intentFilter.addAction(Intent.ACTION_USER_PRESENT);
             context.registerReceiver(new BroadcastReceiver() {
                 @Override
                 public void onReceive(final Context context, Intent intent) {
+                    // Is there a lockscreen still visible to the user?
+                    // If yes, we can't start backgroudn services yet, so we'll need to wait until we get ACTION_USER_PRESENT
                     KeyguardManager myKM = (KeyguardManager) context.getSystemService(Context.KEYGUARD_SERVICE);
                     boolean isPhoneLocked = myKM.inKeyguardRestrictedInputMode();
                     if(!isPhoneLocked) {
                         context.unregisterReceiver(this);
                         final Handler handler = new Handler();
                         // Even when all the stars are aligned,
-                        // starting the service needs to be delayed because the Intent is fired *before* the app is considered in the foreground.
+                        // starting the service needs to be slightly delayed because the Intent is fired *before* the app is considered in the foreground.
                         // Each new release of Android manages to make the developer life harder.
                         // Can't wait for the next one.
                         handler.postDelayed(new Runnable() {
@@ -209,6 +213,8 @@ public class DataHandler extends BroadcastReceiver
                     }
                 }
             }, intentFilter);
+
+            // Stop here for now, the Receiver will re-trigger the whole flow when services can be started.
             return;
         }
 

--- a/app/src/main/java/fr/neamar/kiss/DataHandler.java
+++ b/app/src/main/java/fr/neamar/kiss/DataHandler.java
@@ -1,6 +1,6 @@
 package fr.neamar.kiss;
 
-import android.app.ActivityManager;
+import android.app.KeyguardManager;
 import android.content.BroadcastReceiver;
 import android.content.ComponentName;
 import android.content.Context;
@@ -9,7 +9,6 @@ import android.content.IntentFilter;
 import android.content.ServiceConnection;
 import android.content.SharedPreferences;
 import android.graphics.Bitmap.CompressFormat;
-import android.os.Build;
 import android.os.IBinder;
 import android.preference.PreferenceManager;
 import android.text.TextUtils;
@@ -27,7 +26,6 @@ import java.util.Set;
 
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
-import androidx.annotation.RequiresApi;
 import fr.neamar.kiss.dataprovider.AppProvider;
 import fr.neamar.kiss.dataprovider.ContactsProvider;
 import fr.neamar.kiss.dataprovider.IProvider;
@@ -75,34 +73,6 @@ public class DataHandler extends BroadcastReceiver
         //  to bind to services)
         this.context = context.getApplicationContext();
 
-        if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.P) {
-            init();
-        }
-        else {
-            // After Android Pie, services can't be started in the background
-            // However... KISS is a launcher, and when your device reboots, Android starts the launcher in the background
-            // This is obviously an issue, so we need to wait until screen is really on to do something
-            // https://github.com/Neamar/KISS/issues/1154
-            if (isScreenOn()) {
-                Log.v(TAG, "Screen is on, services can start normally");
-                init();
-            }
-            else {
-                Log.v(TAG, "Screen off, delaying services start.");
-                IntentFilter intentFilter = new IntentFilter(Intent.ACTION_SCREEN_ON);
-                context.registerReceiver(new BroadcastReceiver() {
-                    @Override
-                    public void onReceive(Context context, Intent intent) {
-                        Log.v(TAG, "Screen turned on, starting background services");
-                        init();
-                        context.unregisterReceiver(this);
-                    }
-                }, intentFilter);
-            }
-        }
-    }
-
-    private void init() {
         start = System.currentTimeMillis();
 
         IntentFilter intentFilter = new IntentFilter(MainActivity.LOAD_OVER);
@@ -121,7 +91,7 @@ public class DataHandler extends BroadcastReceiver
         // (this way, we don't need to reload the app list everytime for instance)
         for (String providerName : PROVIDER_NAMES) {
             if (prefs.getBoolean("enable-" + providerName, true)) {
-                this.connectToProvider(providerName);
+                this.connectToProvider(providerName, 0);
             }
         }
 
@@ -143,7 +113,7 @@ public class DataHandler extends BroadcastReceiver
             String providerName = key.substring(7);
             if (PROVIDER_NAMES.contains(providerName)) {
                 if (sharedPreferences.getBoolean(key, true)) {
-                    this.connectToProvider(providerName);
+                    this.connectToProvider(providerName, 0);
                 } else {
                     this.disconnectFromProvider(providerName);
                 }
@@ -179,7 +149,7 @@ public class DataHandler extends BroadcastReceiver
      *
      * @param name Data provider name (i.e.: `ContactsProvider` → `"contacts"`)
      */
-    private void connectToProvider(final String name) {
+    private void connectToProvider(final String name, final int counter) {
         // Do not continue if this provider has already been connected to
         if (this.providers.containsKey(name)) {
             return;
@@ -199,8 +169,36 @@ public class DataHandler extends BroadcastReceiver
             // of the activity
             this.context.startService(intent);
         } catch (IllegalStateException e) {
+            // When KISS is the default launcher,
+            // the system will try to start KISS in the background after a reboot
+            // however at this point we're not allowed to start services, and an IllegalStateException will be thrown
+            // We'll then add a broadcast receiver for the next time the user turns his screen on
+            // (or passes the lockscreen) to retry at this point
             // https://github.com/Neamar/KISS/issues/1130
-            Log.e("KISS", "Unable to start service for " + name + ". This is likely because a broadcast receiver was triggered and KISS is not the default home app, so services are not running and can't be started in this context.");
+            // https://github.com/Neamar/KISS/issues/1154
+            Log.w(TAG, "Unable to start service for " + name + ". KISS is probably not in the foreground. Service will automatically be started when KISS gets to the foreground.");
+
+            if(counter > 4) {
+                Log.e(TAG, "Already tried and failed four times to start service. Giving up.");
+                return;
+            }
+
+            IntentFilter intentFilter = new IntentFilter();
+            intentFilter.addAction(Intent.ACTION_SCREEN_ON);
+            intentFilter.addAction(Intent.ACTION_USER_PRESENT);
+            context.registerReceiver(new BroadcastReceiver() {
+                @Override
+                public void onReceive(Context context, Intent intent) {
+                    Log.i(TAG, "Screen turned on or unlocked, retrying to start background services");
+
+                    KeyguardManager myKM = (KeyguardManager) context.getSystemService(Context.KEYGUARD_SERVICE);
+                    boolean isPhoneLocked = myKM.inKeyguardRestrictedInputMode();
+                    if(!isPhoneLocked) {
+                        connectToProvider(name, counter + 1);
+                        context.unregisterReceiver(this);
+                    }
+                }
+            }, intentFilter);
             return;
         }
 
@@ -679,13 +677,6 @@ public class DataHandler extends BroadcastReceiver
         }
 
         return null;
-    }
-
-    @RequiresApi(api = Build.VERSION_CODES.P)
-    private boolean isScreenOn() {
-        ActivityManager.RunningAppProcessInfo appProcessInfo = new ActivityManager.RunningAppProcessInfo();
-        ActivityManager.getMyMemoryState(appProcessInfo);
-        return (appProcessInfo.importance == ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND || appProcessInfo.importance == ActivityManager.RunningAppProcessInfo.IMPORTANCE_VISIBLE);
     }
 
     public TagsHandler getTagsHandler() {

--- a/app/src/main/java/fr/neamar/kiss/DataHandler.java
+++ b/app/src/main/java/fr/neamar/kiss/DataHandler.java
@@ -8,11 +8,13 @@ import android.content.IntentFilter;
 import android.content.ServiceConnection;
 import android.content.SharedPreferences;
 import android.graphics.Bitmap.CompressFormat;
+import android.hardware.display.DisplayManager;
+import android.os.Build;
 import android.os.IBinder;
-import android.os.PowerManager;
 import android.preference.PreferenceManager;
 import android.text.TextUtils;
 import android.util.Log;
+import android.view.Display;
 
 import java.io.ByteArrayOutputStream;
 import java.util.ArrayList;
@@ -26,6 +28,7 @@ import java.util.Set;
 
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
+import androidx.annotation.RequiresApi;
 import fr.neamar.kiss.dataprovider.AppProvider;
 import fr.neamar.kiss.dataprovider.ContactsProvider;
 import fr.neamar.kiss.dataprovider.IProvider;
@@ -42,8 +45,6 @@ import fr.neamar.kiss.pojo.Pojo;
 import fr.neamar.kiss.pojo.ShortcutsPojo;
 import fr.neamar.kiss.searcher.Searcher;
 import fr.neamar.kiss.utils.UserHandle;
-
-import static android.content.Context.POWER_SERVICE;
 
 public class DataHandler extends BroadcastReceiver
         implements SharedPreferences.OnSharedPreferenceChangeListener {
@@ -82,10 +83,8 @@ public class DataHandler extends BroadcastReceiver
             // After Android Pie, services can't be started in the background
             // However... KISS is a launcher, and when your device reboots, Android starts the launcher in the background
             // This is obviously an issue, so we need to wait until screen is really on to do something
-            // (isInteractive() should be a good enough approximation of isScreenOn() after API 20
             // https://github.com/Neamar/KISS/issues/1154
-            PowerManager powerManager = (PowerManager) context.getSystemService(POWER_SERVICE);
-            if (powerManager.isInteractive()) {
+            if (screenIsOn()) {
                 init();
             }
             else {
@@ -680,6 +679,17 @@ public class DataHandler extends BroadcastReceiver
         }
 
         return null;
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.P)
+    private boolean screenIsOn() {
+        DisplayManager dm = (DisplayManager) context.getSystemService(Context.DISPLAY_SERVICE);
+        for (Display display : dm.getDisplays()) {
+            if (display.getState() == Display.STATE_ON) {
+                return true;
+            }
+        }
+        return false;
     }
 
     public TagsHandler getTagsHandler() {

--- a/app/src/main/java/fr/neamar/kiss/DataHandler.java
+++ b/app/src/main/java/fr/neamar/kiss/DataHandler.java
@@ -9,6 +9,7 @@ import android.content.IntentFilter;
 import android.content.ServiceConnection;
 import android.content.SharedPreferences;
 import android.graphics.Bitmap.CompressFormat;
+import android.os.Handler;
 import android.os.IBinder;
 import android.preference.PreferenceManager;
 import android.text.TextUtils;
@@ -188,13 +189,19 @@ public class DataHandler extends BroadcastReceiver
             intentFilter.addAction(Intent.ACTION_USER_PRESENT);
             context.registerReceiver(new BroadcastReceiver() {
                 @Override
-                public void onReceive(Context context, Intent intent) {
+                public void onReceive(final Context context, Intent intent) {
                     KeyguardManager myKM = (KeyguardManager) context.getSystemService(Context.KEYGUARD_SERVICE);
                     boolean isPhoneLocked = myKM.inKeyguardRestrictedInputMode();
                     if(!isPhoneLocked) {
-                        Log.i(TAG, "Screen turned on or unlocked, retrying to start background services");
-                        connectToProvider(name, counter + 1);
                         context.unregisterReceiver(this);
+                        final Handler handler = new Handler();
+                        handler.postDelayed(new Runnable() {
+                            @Override
+                            public void run() {
+                                Log.i(TAG, "Screen turned on or unlocked, retrying to start background services");
+                                connectToProvider(name, counter + 1);
+                            }
+                        }, 10);
                     }
                 }
             }, intentFilter);

--- a/app/src/main/java/fr/neamar/kiss/DataHandler.java
+++ b/app/src/main/java/fr/neamar/kiss/DataHandler.java
@@ -86,15 +86,16 @@ public class DataHandler extends BroadcastReceiver
                 init();
             }
             else {
+                Log.v(TAG, "Screen off, delaying services start.");
                 IntentFilter intentFilter = new IntentFilter(Intent.ACTION_SCREEN_ON);
-                final BroadcastReceiver screenOnReceiver = new BroadcastReceiver() {
+                context.registerReceiver(new BroadcastReceiver() {
                     @Override
                     public void onReceive(Context context, Intent intent) {
+                        Log.v(TAG, "Screen turned on, starting background services");
                         init();
-                        context.unregisterReceiver(screenOnReceiver);
+                        context.unregisterReceiver(this);
                     }
-                };
-                context.registerReceiver(screenOnReceiver, intentFilter);
+                }, intentFilter);
             }
         }
     }

--- a/app/src/main/java/fr/neamar/kiss/DataHandler.java
+++ b/app/src/main/java/fr/neamar/kiss/DataHandler.java
@@ -195,6 +195,10 @@ public class DataHandler extends BroadcastReceiver
                     if(!isPhoneLocked) {
                         context.unregisterReceiver(this);
                         final Handler handler = new Handler();
+                        // Even when all the stars are aligned,
+                        // starting the service needs to be delayed because the Intent is fired *before* the app is considered in the foreground.
+                        // Each new release of Android manages to make the developer life harder.
+                        // Can't wait for the next one.
                         handler.postDelayed(new Runnable() {
                             @Override
                             public void run() {

--- a/app/src/main/java/fr/neamar/kiss/DataHandler.java
+++ b/app/src/main/java/fr/neamar/kiss/DataHandler.java
@@ -178,8 +178,8 @@ public class DataHandler extends BroadcastReceiver
             // https://github.com/Neamar/KISS/issues/1154
             Log.w(TAG, "Unable to start service for " + name + ". KISS is probably not in the foreground. Service will automatically be started when KISS gets to the foreground.");
 
-            if(counter > 4) {
-                Log.e(TAG, "Already tried and failed four times to start service. Giving up.");
+            if(counter > 20) {
+                Log.e(TAG, "Already tried and failed twenty times to start service. Giving up.");
                 return;
             }
 
@@ -189,11 +189,10 @@ public class DataHandler extends BroadcastReceiver
             context.registerReceiver(new BroadcastReceiver() {
                 @Override
                 public void onReceive(Context context, Intent intent) {
-                    Log.i(TAG, "Screen turned on or unlocked, retrying to start background services");
-
                     KeyguardManager myKM = (KeyguardManager) context.getSystemService(Context.KEYGUARD_SERVICE);
                     boolean isPhoneLocked = myKM.inKeyguardRestrictedInputMode();
                     if(!isPhoneLocked) {
+                        Log.i(TAG, "Screen turned on or unlocked, retrying to start background services");
                         connectToProvider(name, counter + 1);
                         context.unregisterReceiver(this);
                     }

--- a/app/src/main/java/fr/neamar/kiss/DataHandler.java
+++ b/app/src/main/java/fr/neamar/kiss/DataHandler.java
@@ -82,7 +82,10 @@ public class DataHandler extends BroadcastReceiver
             // After Android Pie, services can't be started in the background
             // However... KISS is a launcher, and when your device reboots, Android starts the launcher in the background
             // This is obviously an issue, so we need to wait until screen is really on to do something
-            if (isScreenOn()) {
+            // (isInteractive() should be a good enough approximation of isScreenOn() after API 20
+            // https://github.com/Neamar/KISS/issues/1154
+            PowerManager powerManager = (PowerManager) context.getSystemService(POWER_SERVICE);
+            if (powerManager.isInteractive()) {
                 init();
             }
             else {
@@ -102,8 +105,6 @@ public class DataHandler extends BroadcastReceiver
 
     private void init() {
         start = System.currentTimeMillis();
-
-
 
         IntentFilter intentFilter = new IntentFilter(MainActivity.LOAD_OVER);
         this.context.getApplicationContext().registerReceiver(this, intentFilter);
@@ -135,11 +136,6 @@ public class DataHandler extends BroadcastReceiver
         ProviderEntry phoneEntry = new ProviderEntry();
         phoneEntry.provider = new PhoneProvider(context);
         this.providers.put("phone", phoneEntry);
-    }
-
-    public boolean isScreenOn() {
-        PowerManager powerManager = (PowerManager) context.getSystemService(POWER_SERVICE);
-        return powerManager.isScreenOn();
     }
 
     @Override

--- a/app/src/main/java/fr/neamar/kiss/DataHandler.java
+++ b/app/src/main/java/fr/neamar/kiss/DataHandler.java
@@ -9,6 +9,7 @@ import android.content.ServiceConnection;
 import android.content.SharedPreferences;
 import android.graphics.Bitmap.CompressFormat;
 import android.os.IBinder;
+import android.os.PowerManager;
 import android.preference.PreferenceManager;
 import android.text.TextUtils;
 import android.util.Log;
@@ -42,6 +43,8 @@ import fr.neamar.kiss.pojo.ShortcutsPojo;
 import fr.neamar.kiss.searcher.Searcher;
 import fr.neamar.kiss.utils.UserHandle;
 
+import static android.content.Context.POWER_SERVICE;
+
 public class DataHandler extends BroadcastReceiver
         implements SharedPreferences.OnSharedPreferenceChangeListener {
     final static private String TAG = "DataHandler";
@@ -67,12 +70,39 @@ public class DataHandler extends BroadcastReceiver
      * Initialize all providers
      */
     public DataHandler(Context context) {
-        start = System.currentTimeMillis();
-
-        // Make sure we are in the context of the main activity
+        // Make sure we are in the context of the main application
         // (otherwise we might receive an exception about broadcast listeners not being able
         //  to bind to services)
         this.context = context.getApplicationContext();
+
+        if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.P) {
+            init();
+        }
+        else {
+            // After Android Pie, services can't be started in the background
+            // However... KISS is a launcher, and when your device reboots, Android starts the launcher in the background
+            // This is obviously an issue, so we need to wait until screen is really on to do something
+            if (isScreenOn()) {
+                init();
+            }
+            else {
+                IntentFilter intentFilter = new IntentFilter(Intent.ACTION_SCREEN_ON);
+                final BroadcastReceiver screenOnReceiver = new BroadcastReceiver() {
+                    @Override
+                    public void onReceive(Context context, Intent intent) {
+                        init();
+                        context.unregisterReceiver(screenOnReceiver);
+                    }
+                };
+                context.registerReceiver(screenOnReceiver, intentFilter);
+            }
+        }
+    }
+
+    private void init() {
+        start = System.currentTimeMillis();
+
+
 
         IntentFilter intentFilter = new IntentFilter(MainActivity.LOAD_OVER);
         this.context.getApplicationContext().registerReceiver(this, intentFilter);
@@ -104,6 +134,11 @@ public class DataHandler extends BroadcastReceiver
         ProviderEntry phoneEntry = new ProviderEntry();
         phoneEntry.provider = new PhoneProvider(context);
         this.providers.put("phone", phoneEntry);
+    }
+
+    public boolean isScreenOn() {
+        PowerManager powerManager = (PowerManager) context.getSystemService(POWER_SERVICE);
+        return powerManager.isScreenOn();
     }
 
     @Override

--- a/app/src/main/java/fr/neamar/kiss/MainActivity.java
+++ b/app/src/main/java/fr/neamar/kiss/MainActivity.java
@@ -12,6 +12,7 @@ import android.content.SharedPreferences;
 import android.database.DataSetObserver;
 import android.os.Build;
 import android.os.Bundle;
+import android.os.PowerManager;
 import android.preference.PreferenceManager;
 import android.text.Editable;
 import android.text.TextWatcher;
@@ -283,9 +284,9 @@ public class MainActivity extends Activity implements QueryInterface, KeyboardSc
         // Fixes bug when dropping onto a textEdit widget which can cause a NPE
         // This fix should be on ALL TextEdit Widgets !!!
         // See : https://stackoverflow.com/a/23483957
-        searchEditText.setOnDragListener( new View.OnDragListener() {
+        searchEditText.setOnDragListener(new View.OnDragListener() {
             @Override
-            public boolean onDrag( View v, DragEvent event) {
+            public boolean onDrag(View v, DragEvent event) {
                 return true;
             }
         });
@@ -655,7 +656,7 @@ public class MainActivity extends Activity implements QueryInterface, KeyboardSc
                     });
                     anim.setDuration(animationDuration);
                     anim.start();
-                } catch(IllegalStateException e) {
+                } catch (IllegalStateException e) {
                     // If the view hasn't been laid out yet, we can't animate it
                     kissBar.setVisibility(View.GONE);
                 }
@@ -811,15 +812,14 @@ public class MainActivity extends Activity implements QueryInterface, KeyboardSc
             mPopup.dismiss();
     }
 
-    public void showMatchingTags( String tag ) {
+    public void showMatchingTags(String tag) {
         runTask(new TagsSearcher(this, tag));
 
         clearButton.setVisibility(View.VISIBLE);
         menuButton.setVisibility(View.INVISIBLE);
     }
 
-    public void showUntagged()
-    {
+    public void showUntagged() {
         runTask(new UntaggedSearcher(this));
 
         clearButton.setVisibility(View.VISIBLE);

--- a/app/src/main/java/fr/neamar/kiss/result/Result.java
+++ b/app/src/main/java/fr/neamar/kiss/result/Result.java
@@ -6,8 +6,6 @@ import android.graphics.Color;
 import android.graphics.drawable.Drawable;
 import android.os.AsyncTask;
 import android.preference.PreferenceManager;
-import androidx.annotation.NonNull;
-import androidx.annotation.StringRes;
 import android.text.Spannable;
 import android.text.SpannableString;
 import android.text.style.ForegroundColorSpan;
@@ -24,6 +22,8 @@ import android.widget.Toast;
 import java.lang.ref.WeakReference;
 import java.util.List;
 
+import androidx.annotation.NonNull;
+import androidx.annotation.StringRes;
 import fr.neamar.kiss.BuildConfig;
 import fr.neamar.kiss.KissApplication;
 import fr.neamar.kiss.MainActivity;
@@ -128,12 +128,17 @@ public abstract class Result {
     }
 
     public String getSection() {
-        // get the normalized first letter of the pojo
-        // Ensure accented characters are never displayed. (É => E)
-        String ch = Character.toString((char) pojo.normalizedName.codePoints[0]);
-        // convert to uppercase otherwise lowercase a -z will be sorted
-        // after upper A-Z
-        return ch.toUpperCase();
+        try {
+            // get the normalized first letter of the pojo
+            // Ensure accented characters are never displayed. (É => E)
+            String ch = Character.toString((char) pojo.normalizedName.codePoints[0]);
+            // convert to uppercase otherwise lowercase a -z will be sorted
+            // after upper A-Z
+            return ch.toUpperCase();
+        } catch(ArrayIndexOutOfBoundsException e) {
+            // Normalized name is empty.
+            return "-";
+        }
     }
 
     /**

